alf/config_util_test.py

```diff
@@ -0,0 +1,148 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from absl import logging
+import pprint
+import alf
+
+
+@alf.configurable
+def test(a, b=123):
+    return a, b
+
+
+@alf.configurable(blacklist=['b'])
+def test_func(a, b=100, c=200):
+    return a, b, c
+
+
+@alf.configurable(blacklist=['b'])
+def test_func2(a, b=100, c=200):
+    return a, b, c
+
+
+@alf.configurable(blacklist=['b'])
+def test_func3(a, b=100, c=200):
+    return a, b, c
+
+
+@alf.configurable("FancyTest")
+def test_func4(arg=10):
+    return arg
+
+
+@alf.configurable
+class Test(object):
+    def __init__(self, a, b, c=10):
+        self._a = a
+        self._b = b
+        self._c = c
+
+    @alf.configurable(whitelist=['c'])
+    def func(self, a, b=10, c=100):
+        return a, b, c
+
+    def __call__(self):
+        return self._a, self._b, self._c
+
+
+@alf.configurable
+class Test2(Test):
+    def __init__(self, a, b):
+        super().__init__(a, b, 5)
+
+    def func(self, a):
+        return a
+
+    def __call__(self):
+        return self._a + 1, self._b + 2, self._c + 3
+
+
+@alf.configurable
+class Test3(Test):
+    def func(self, a):
+        return a
+
+    def __call__(self):
+        return self._a - 1, self._b - 2, self._c - 3
+
+
+class ConfigTest(alf.test.TestCase):
+    def test_config1(self):
+
+        # Test simple function
+        self.assertEqual(test(1), (1, 123))
+        alf.config({'test.a': 3})
+        self.assertEqual(test(), (3, 123))
+        # Test config after the value has been used
+        self.assertRaisesRegex(ValueError, "test.b' has already been used",
+                               alf.config, {'test.b': None})
+
+        # Test class
+        obj = Test(1, 2)
+        self.assertEqual(obj(), (1, 2, 10))
+        alf.config({'Test.b': 5})
+        obj = Test(1)
+        self.assertEqual(obj(), (1, 5, 10))
+
+        self.assertRaises(ValueError, alf.config1, 'c', 32)
+
+        # Test class member function
+        # Test whitelist
+        alf.config({'Test.func.c': 30})
+        # Test whitelist
+        self.assertRaises(ValueError, alf.config1, 'Test.func.b', 32)
+        self.assertEqual(obj.func(1, 2, 3), (1, 2, 3))
+        self.assertEqual(obj.func(3, 5), (3, 5, 30))
+        self.assertRaisesRegex(TypeError, "missing 1 required positional",
+                               obj.func)
+
+        # Test blacklist
+        self.assertRaises(ValueError, alf.config1, 'test_func.b', 30)
+        alf.config1('test_func.c', 15)
+        self.assertEqual(test_func(10, 20), (10, 20, 15))
+
+        alf.config("FancyTest", arg=3)
+        self.assertEqual(test_func4(), 3)
+
+        # Test duplicated config
+        with self.assertLogs() as ctx:
+            alf.config1('test_func2.c', 15)
+            alf.config1('test_func2.c', 16)
```

Comment:
Can command line --conf_param override existing configs in the conf file?
My personal habit is to have the conf file being self sufficient (so it can be used standalone), and have conf_param provide ways of customizing configs based on the conf file (e.g. running one off experiments tweaking some configs).
Does that make sense?

Reply:
Then I don't quite understand why some confs in the conf file may depend on --conf_param values?